0025-nat-traversal.adoc

```diff
@@ -0,0 +1,174 @@
+Bob notifies Alice that he can reached through the aforementioned publicly facing port and IP address.
+
+This solution does not require any changes to the COMIT codebase and places the burden on the user to configure their NAT.
+Users who do not have control over the configuration of their NAT can not use this method.
+We cannot completely rely on port forwarding because users on mobile, shared or corporate managed networks will not have access to NAT configuration.
+
+- + Alice and Bob can communicate through a direct connection
+
+- - Requires Bob to have access to NAT configuration
+
+- - Requires Bob to configure his NAT
+
+
+=== Automatic Port Forwarding using UPnP
+
+A UPnP enabled NAT automatically forwards publicly facing ports as requested by an application.
+Unlike manual port forwarding, Bob is not required to configure his NAT.
+Not all NATs are UPnP capable and it is often disabled or not used by peer to peer applications.
+Bitcoin core removed UPnP support due to security concerns: https://bitcoin.org/en/alert/2015-10-12-upnp-vulnerability.
+
+- + Does not require Bob to configure his NAT
+
+- + Alice and Bob can communicate through a direct connection
+
+- - Requires the user to have a UPnP capable and enabled NAT (UPnP capable routers are fairly common in home networks)
+
+- - Opens up the user to security vulnerabilities arising from the NAT blindly trusting applications and allowing to open externally facing ports
+
+=== Reverse connection
+
+Instead of Alice attempting to reach Bob through his NAT. Bob dials Alice and Alice uses the connection to send to data to Bob.
+This technique only works if Alice knows her public IP and port and has provided it to Bo.
+Since Alice is unable to dial Bob, Alice is unable to initiate communication and must wait to be dialled by a Bob.
+
+Reverse connections cannot be used because many of our protocols require both peers to be able to initate a connection with each other.
+This means that COMIT protocols will have to redesigned to use the connection initiated by dialer to complete an atomic swap.
+Peers behind a NAT will be unable to play an active role in the COMIT network.
+
+- + Does not require Bob to configure his NAT
+
+- + Does not require the user to have access to NAT configuration
+
+- + Alice and Bob can  communicate through a direct connection
+
+- - Alice can not dial Bob and therefore cannot initiate a swap or use Bob as a peer to propagate information to the network
+
+- - We will most likely need to redesign protocols to utilise reverse connections
+
+- - Requires Bob to have a known external IP
+
+=== STUN/Hole Punching
+
+Bob sends UDP messages to a third peer known as a STUN server.
+The STUN server will return which ports and public IP Bobs's UDP messages originated from.
+Bob has now discovered public IP and ports which peers can use to dial him.
+Bob can provide this knowledge to Alice, allowing Alice to dial him.
```

Comment:
Not entirely true, even with an IP and port, if this is behind NAT the port will need to be forwarded to the local machine's IP address.  Might be worth mentioning this quote from wikipedia
> STUN is not a self-contained NAT traversal solution applicable in all NAT deployment scenarios and does not work correctly with all of them. It is a tool among other methods and it is a tool for other protocols in dealing with NAT traversal, most notably Traversal Using Relay NAT (TURN) and Interactive Connectivity Establishment (ICE).